nautobot/extras/plugins/__init__.py

```diff
@@ -491,6 +518,106 @@ def register_filter_extensions(filter_extensions, plugin_name):
+    """
+
+    model = None
+    table_columns = {}
+    add_to_default_columns = ()
+    remove_from_default_columns = ()
+
+    @classmethod
+    def alter_queryset(cls, queryset):
+        """Alter the View class QuerySet.
+
+        This is a good place to add `prefetch_related` to the view queryset.
+        example:
+            return queryset.prefetch_related("my_model_set")
+        """
+        return queryset
+
+
+def register_table_extensions(table_extensions, app_name):
+    """Register a list of TableExtension classes."""
+    for table_extension in table_extensions:
+        _validate_is_subclass_of_table_extension(table_extension)
+        _register_table_base_columns(table_extension, app_name)
+        _modify_default_table_columns(table_extension)
+        _alter_table_view_queryset(table_extension, app_name)
+
+
+def _register_table_base_columns(table_extension, app_name):
+    """Register each column in the TableExtension."""
+    from nautobot.core.utils.lookup import get_table_for_model
+
+    table = get_table_for_model(table_extension.model)
+    for name, column in table_extension.table_columns.items():
+        _validate_table_column_name_is_prefixed_with_app_name(name, app_name)
+        try:
+            _add_column_to_table_base_columns(table, name, column)
+        except AttributeError as error:
+            logger.error(*error.args)
+
+
+def _add_column_to_table_base_columns(table, column_name, column):
+    """Attach a column to an existing table."""
+    import django_tables2
+
+    if not isinstance(column, django_tables2.Column):
+        raise TypeError(f"Custom column `{column_name}` is not an instance of django_tables2.Column.")
+
+    if column_name in table.base_columns:
+        raise AttributeError(f"There was a conflict with table column `{column_name}`, the custom column was ignored.")
+
+    table.base_columns[column_name] = column
+
+
+def _alter_table_view_queryset(table_extension, app_name):
+    """Replace the model view queryset with an optimized queryset from the app."""
+    from nautobot.core.utils.lookup import get_view_for_model
+
+    view = get_view_for_model(table_extension.model, view_type="List")
+    view.queryset = table_extension.alter_queryset(view.queryset)
+
+
+def _modify_default_table_columns(table_extension):
+    """Add or remove columns from the table default columns."""
+    from nautobot.core.utils.lookup import get_table_for_model
+
+    table = get_table_for_model(table_extension.model)
+    add_registered_columns = [name for name in table_extension.add_to_default_columns if name in table.base_columns]
+
+    if add_registered_columns:
+        table.Meta.default_columns = (*table.Meta.default_columns, *add_registered_columns)
+
+    if table_extension.remove_from_default_columns:
+        table.Meta.default_columns = tuple(
+            column for column in table.Meta.default_columns if column not in table_extension.remove_from_default_columns
+        )
```

Comment:
It *might* be nice to log a debug message if `remove_from_default_columns` includes any items that aren't actually in the table's default_columns? Definitely no higher than `debug` though.

Reply:
I've refactored this code and added debug logging for when a column name is missing.